libs/parallelism/algorithms/tests/unit/container_algorithms/rotate_copy_range.cpp

```diff
@@ -19,6 +20,122 @@
 #include "test_utils.hpp"
 
 ///////////////////////////////////////////////////////////////////////////////
+void test_rotate_copy_sent(typename IteratorTag)
+{
+    typedef std::vector<std::size_t>::iterator base_iterator;
+    typedef test::test_iterator<base_iterator, IteratorTag> iterator;
+
+    typedef test::test_container<std::vector<std::size_t>, IteratorTag>
+        test_vector;
+
+    test_vector c(10007);
```

Comment:
For sentinel tests I prefer using a normal std::vector instead of the test_iterator and test_container because that has issues with the sentinels.

Reply:
I will rewrite it and back to you. Thanks

Reply:
No idea, I guess we can make it work, I never really looked into it. Feel free to experiment and make it work with the test_vector

Reply:
> No idea, I guess we can make it work, I never really looked into it. Feel free to experiment and make it work with the test_vector

got it. will do.

Reply:
Hi Akhil, I am not sure if the iter_sent test is fine, maybe it's too complicated, so could you please take a look at this link.

[https://gist.github.com/hcq9102/59cbb237d19e7a6f3b59a5e9b69081c6 ](url)

Reply:
`// test case for iterator - sentinel_value
template <typename IteratorTag>
void test_rotate_sent(IteratorTag)
{
    std::size_t const size = 10007;
    std::vector<std::size_t> c(size), d1;

    std::iota(std::begin(c), std::end(c), std::rand());
    std::copy(std::begin(c), std::end(c), std::back_inserter(d1));

    std::size_t mid_pos = std::rand() % (c.size()-1);    //-V104
    auto mid = std::begin(c);
    std::advance(mid, mid_pos);

    hpx::ranges::rotate(std::begin(c), iterator(mid), sentinel<std::size_t>{*(std::end(c) - 1)});

    auto mid1 = std::begin(d1);
    std::advance(mid1, mid_pos);
    std::rotate(std::begin(d1), mid1, std::end(d1) - 1);

    std::size_t count = 0;
    HPX_TEST(std::equal(std::begin(c), sentinel<std::size_t>{*(std::end(c) - 1)}, 
        std::begin(d1), [&count](std::size_t v1, std::size_t v2) -> bool {
            HPX_TEST_EQ(v1, v2);
            ++count;
            return v1 == v2;
        }));
    HPX_TEST_EQ(count, d1.size()-1);  // d1.size()   ----> d1.size()-1 ?
}
`

Reply:
`// test case for iterator - sentinel_value
template
void test_rotate_sent(IteratorTag)
{
std::size_t const size = 10007;
std::vectorstd::size_t c(size), d1;

std::iota(std::begin(c), std::end(c), std::rand());
std::copy(std::begin(c), std::end(c), std::back_inserter(d1));

std::size_t mid_pos = std::rand() % (c.size()-1);    //-V104
auto mid = std::begin(c);
std::advance(mid, mid_pos);

hpx::ranges::rotate(std::begin(c), iterator(mid), sentinel<std::size_t>{*(std::end(c) - 1)});

auto mid1 = std::begin(d1);
std::advance(mid1, mid_pos);
std::rotate(std::begin(d1), mid1, std::end(d1) - 1);

std::size_t count = 0;
HPX_TEST(std::equal(std::begin(c), sentinel<std::size_t>{*(std::end(c) - 1)}, 
    std::begin(d1), [&count](std::size_t v1, std::size_t v2) -> bool {
        HPX_TEST_EQ(v1, v2);
        ++count;
        return v1 == v2;
    }));
HPX_TEST_EQ(count, d1.size()-1);  // d1.size()   ----> d1.size()-1 ?
}
`

Reply:
`// test case for iterator - sentinel_value
template <typename IteratorTag>
void test_rotate_sent(IteratorTag)
{
std::size_t const size = 10007;
std::vectorstd::size_t c(size), d1;

std::iota(std::begin(c), std::end(c), std::rand());
std::copy(std::begin(c), std::end(c), std::back_inserter(d1));

std::size_t mid_pos = std::rand() % (c.size()-1); //-V104
auto mid = std::begin(c);
std::advance(mid, mid_pos);

hpx::ranges::rotate(std::begin(c), iterator(mid), sentinelstd::size_t{*(std::end(c) - 1)});

auto mid1 = std::begin(d1);
std::advance(mid1, mid_pos);
std::rotate(std::begin(d1), mid1, std::end(d1) - 1);

std::size_t count = 0;
HPX_TEST(std::equal(std::begin(c), sentinelstd::size_t{*(std::end(c) - 1)},
std::begin(d1), [&count](std::size_t v1, std::size_t v2) -> bool {
HPX_TEST_EQ(v1, v2);
++count;
return v1 == v2;
}));
HPX_TEST_EQ(count, d1.size()-1); // d1.size() ----> d1.size()-1 ?
}`

Reply:
part of the code of the invalid link like above... ( ...test.../rotate_range.cpp); 
this line :  `HPX_TEST_EQ(count, d1.size()-1); // d1.size() ----> d1.size()-1 ?`
it should be d1.size() ?

Thanks a lot.

Reply:
// test case for iterator - sentinel_value
template
void test_rotate_sent(IteratorTag)
{
std::size_t const size = 10007;
std::vectorstd::size_t c(size), d1;

std::iota(std::begin(c), std::end(c), std::rand());
std::copy(std::begin(c), std::end(c), std::back_inserter(d1));

std::size_t mid_pos = std::rand() % (c.size()-1); //-V104
auto mid = std::begin(c);
std::advance(mid, mid_pos);

hpx::ranges::rotate(std::begin(c), iterator(mid), sentinelstd::size_t{*(std::end(c) - 1)});

auto mid1 = std::begin(d1);
std::advance(mid1, mid_pos);
std::rotate(std::begin(d1), mid1, std::end(d1) - 1);

std::size_t count = 0;
HPX_TEST(std::equal(std::begin(c), std::end(c) - 1,
std::begin(d1), [&count](std::size_t v1, std::size_t v2) -> bool {
HPX_TEST_EQ(v1, v2);
++count;
return v1 == v2;
}));
HPX_TEST_EQ(count, d1.size()-1); // d1.size() ----> d1.size()-1 ?
}`

part of the code of the invalid link like above... ( ...test.../rotate_range.cpp);
this line : `HPX_TEST_EQ(count, d1.size()-1); // d1.size() ----> d1.size()-1 ?`
it should be d1.size() ?

Thanks a lot.

Reply:
`// test case for iterator - sentinel_value
template
void test_rotate_sent(IteratorTag)
{
std::size_t const size = 10007;
std::vectorstd::size_t c(size), d1;

std::iota(std::begin(c), std::end(c), std::rand());
std::copy(std::begin(c), std::end(c), std::back_inserter(d1));

std::size_t mid_pos = std::rand() % (c.size()-1); //-V104
auto mid = std::begin(c);
std::advance(mid, mid_pos);

hpx::ranges::rotate(std::begin(c), iterator(mid), sentinelstd::size_t{*(std::end(c) - 1)});

auto mid1 = std::begin(d1);
std::advance(mid1, mid_pos);
std::rotate(std::begin(d1), mid1, std::end(d1) - 1);

std::size_t count = 0;
HPX_TEST(std::equal(std::begin(c), std::end(c) - 1,
std::begin(d1), [&count](std::size_t v1, std::size_t v2) -> bool {
HPX_TEST_EQ(v1, v2);
++count;
return v1 == v2;
}));
HPX_TEST_EQ(count, d1.size()-1); // d1.size() ----> d1.size()-1 ?
}`
------------------------------------------------------------------------------------------------------------
part of the code of the invalid link like above... ( .../test/.../rotate_range.cpp);
this line : `HPX_TEST_EQ(count, d1.size()-1); // d1.size() ----> d1.size()-1 ?`
it should be d1.size() ?

Thanks a lot.

Reply:
// test case for iterator - sentinel_value
template<typename IteratorTag>
void test_rotate_sent(IteratorTag)
{
      std::size_t const size = 10007;
      std::vectorstd::size_t c(size), d1;
      
      std::iota(std::begin(c), std::end(c), std::rand());
      std::copy(std::begin(c), std::end(c), std::back_inserter(d1));
      
      std::size_t mid_pos = std::rand() % (c.size()-1); //-V104
      auto mid = std::begin(c);
      std::advance(mid, mid_pos);
      
      hpx::ranges::rotate(std::begin(c), iterator(mid), sentinelstd::size_t{*(std::end(c) - 1)});
      
      auto mid1 = std::begin(d1);
      std::advance(mid1, mid_pos);
      std::rotate(std::begin(d1), mid1, std::end(d1) - 1);
      
      std::size_t count = 0;
      HPX_TEST(std::equal(std::begin(c), std::end(c) - 1,
      std::begin(d1), [&count](std::size_t v1, std::size_t v2) -> bool {
      HPX_TEST_EQ(v1, v2);
      ++count;
      return v1 == v2;
      }));
      HPX_TEST_EQ(count, d1.size()-1); // d1.size() ----> d1.size()-1 ?
}
-----------------------------------------------------------
part of the code of the invalid link like above... ( .../test/.../rotate_range.cpp);
this line : `HPX_TEST_EQ(count, d1.size()-1); // d1.size() ----> d1.size()-1 ?`
it should be d1.size() ?

Thanks a lot.

Reply:
// test case for iterator - sentinel_value
template <typename IteratorTag >
void test_rotate_sent(IteratorTag)
{
	std::size_t const size = 10007;
	std::vectorstd::size_t c(size), d1;

	 std::iota(std::begin(c), std::end(c), std::rand());
	 std::copy(std::begin(c), std::end(c), std::back_inserter(d1));
	  
	 std::size_t mid_pos = std::rand() % (c.size()-1); //-V104
	 auto mid = std::begin(c);
	 std::advance(mid, mid_pos);
	  
	 hpx::ranges::rotate(std::begin(c), iterator(mid), sentinelstd::size_t{*(std::end(c) - 1)});
	  
	 auto mid1 = std::begin(d1);
	 std::advance(mid1, mid_pos);
	 std::rotate(std::begin(d1), mid1, std::end(d1) - 1);
	  
	 std::size_t count = 0;
	 HPX_TEST(std::equal(std::begin(c), std::end(c) - 1,
	 std::begin(d1), [&count](std::size_t v1, std::size_t v2) -> bool {
	 HPX_TEST_EQ(v1, v2);
	  ++count;
	  return v1 == v2;
	 }));
	 HPX_TEST_EQ(count, d1.size()-1); // d1.size() ----> d1.size()-1 ?
}

-------------------------------------------------------------------------------------
part of the code of the invalid link like above... ( .../test/.../rotate_range.cpp);
this line : `HPX_TEST_EQ(count, d1.size()-1); // d1.size() ----> d1.size()-1 ?`
it should be d1.size() ?

Thanks a lot.

Reply:
template <typename IteratorTag>
void test_rotate_sent(IteratorTag)
{
    std::vector<std::size_t> c(10007);
    std::vector<std::size_t> d1(c.size());

    std::iota(std::begin(c), std::end(c), std::rand());
    std::copy(std::begin(c), std::end(c), std::back_inserter(d1));

    std::size_t mid_pos = std::rand() % (c.size() - 1);    //-V104

    auto mid = std::begin(c);
    std::advance(mid, mid_pos);

    hpx::ranges::rotate(std::begin(c), iterator(mid),
        sentinel<std::size_t>{*(std::end(c) - 1)});

    auto mid1 = std::begin(d1);
    std::advance(mid1, mid_pos);
    std::rotate(std::begin(d1), mid1, std::end(d1) - 1);

    std::size_t count = 0;
    HPX_TEST(std::equal(std::begin(c), std::end(c) - 1, std::begin(d1),
        [&count](std::size_t v1, std::size_t v2) -> bool {
            HPX_TEST_EQ(v1, v2);
            ++count;
            return v1 == v2;
        }));
    HPX_TEST_EQ(count, d1.size() - 1);  // d1.size() ----> d1.size()-1 ?
}

Reply:
> template <typename IteratorTag>
void test_rotate_sent(IteratorTag)
{
    std::vector<std::size_t> c(10007);
    std::vector<std::size_t> d1(c.size());

    std::iota(std::begin(c), std::end(c), std::rand());
    std::copy(std::begin(c), std::end(c), std::back_inserter(d1));

    std::size_t mid_pos = std::rand() % (c.size()-1);    //-V104

    auto mid = std::begin(c);
    std::advance(mid, mid_pos);

    hpx::ranges::rotate(std::begin(c), iterator(mid),
        sentinel<std::size_t>{*(std::end(c) - 1)});

    auto mid1 = std::begin(d1);
    std::advance(mid1, mid_pos);
    std::rotate(std::begin(d1), mid1, std::end(d1) - 1);

    std::size_t count = 0;
    HPX_TEST(std::equal(std::begin(c), std::end(c) - 1, std::begin(d1),
        [&count](std::size_t v1, std::size_t v2) -> bool {
            HPX_TEST_EQ(v1, v2);
            ++count;
            return v1 == v2;
        }));
    HPX_TEST_EQ(count, d1.size());
}

Reply:
[https://gist.github.com/hcq9102/e59a900d0f18913c9d1314a17cc13a60](url)

part of the code of the invalid link like above... ( .../test/.../rotate_range.cpp);
this line : HPX_TEST_EQ(count, d1.size()-1); // d1.size() ----> d1.size()-1 ?
it should be d1.size() ?

Thanks a lot.

Reply:
https://gist.github.com/hcq9102/e59a900d0f18913c9d1314a17cc13a60

part of the code of the invalid link like above... ( .../test/.../rotate_range.cpp);
this line : HPX_TEST_EQ(count, d1.size()-1); // d1.size() ----> d1.size()-1 ?
it should be d1.size() ?

Thanks a lot.

Reply:
https://gist.github.com/hcq9102/e59a900d0f18913c9d1314a17cc13a60

part of the code of the invalid link like above... ( .../test/.../rotate_range.cpp);
the end line : `HPX_TEST_EQ(count, d1.size()-1); // d1.size() ----> d1.size()-1 ?`
it should be d1.size() ?

Thanks a lot.